### PR TITLE
Jcouls29/dynamic

### PR DESCRIPTION
**Pull request by Jcouls29**
Due to a need for a more dynamic setup of workflow, a static method was added to AutomatonymousStateMachine to allow for the builder pattern. New interfaces were created and all current tests were copied and converted to the builder pattern to ensure everything was up to spec.

The general principle is that, by using the builder pattern, anyone can build a system (or framework) on top of this library where the need to build the workflow within the constructor isn't ideal. By allowing the states and events to live outside of the workflow, allows for more flexibility for these type of use cases, especially in a dynamic workflow situation where configurable workflows may be necessary.

Open to any thoughts on this approach.

**Example Usage**
```csharp
State Waiting;
Event Start;
Event First;
Event Second;
Event Third;

AutomatonymousStateMachine<Instance>
    .Build(builder => builder
        .State("Waiting", out Waiting)
        .Event("Start", out Start)
        .Event("First", out First)
        .Event("Second", out Second)
        .Event("Third", out Third)
        .Initially()
            .When(Start, b => b.TransitionTo(Waiting))
        .During(Waiting)
            .When(First, b => b.Then(context => { context.Instance.CalledAfterAll = false; }))
            .When(Second, b => b.Then(context => { context.Instance.CalledAfterAll = false; }))
        .CompositeEvent(Third, b => b.CompositeStatus, First, Second)
        .During(Waiting)
            .When(Third, b => b
                .Then(context =>
                {
                    context.Instance.Called = true;
                    context.Instance.CalledAfterAll = true;
                })
                .Finalize()
            )

    );
```

### CODE REVIEW

File: src/Automatonymous/AutomatonymousStateMachine.cs [hunk 1]
```diff
@@ -244,13 +247,20 @@ protected void Name(string machineName)
...
         /// <param name="propertyExpression"></param>
-        protected virtual void Event(Expression<Func<Event>> propertyExpression)
+        protected internal virtual void Event(Expression<Func<Event>> propertyExpression)
         {
             PropertyInfo property = propertyExpression.GetPropertyInfo();
 
             DeclareTriggerEvent(property);
         }
 
+        protected internal virtual Event Event(string name)
```
Comment (phatboyg) [inline src/Automatonymous/AutomatonymousStateMachine.cs:257]: Would be good to dry these up with `DeclareDataEvent`, and use the property initializer only where needed.

File: src/Automatonymous/AutomatonymousStateMachine.cs [hunk 2]
```diff
@@ -431,17 +448,144 @@ void DeclareDataEvent<T>(PropertyInfo property)
             }
         }
 
+        protected internal virtual void CompositeEvent(string name,
```
Comment (phatboyg) [inline src/Automatonymous/AutomatonymousStateMachine.cs:451]: These will need to just be internal, I don't want them showing up as methods in state machine being declared in the constructor.

File: src/Automatonymous/AutomatonymousStateMachine.cs [hunk 3]
```diff
@@ -533,13 +677,43 @@ protected virtual void SubState(Expression<Func<State>> propertyExpression, Stat
...
+
+        protected internal virtual void SubState(State<TInstance> subState, State<TInstance> superState)
+        {
+            if (superState == null)
+                throw new ArgumentNullException(nameof(superState));
+
+            State<TInstance> superStateInstance = GetState(superState.Name);
+
+            // If the state was already defined, don't define it again
+            superStateInstance.AddSubstate(subState);
```
Comment (phatboyg) [inline src/Automatonymous/AutomatonymousStateMachine.cs:707]: Why is this needed? If the state is already created, it should have the superState already from the constructor. It isn't adding it to a list, Add is confusing, and this method seems unnecessary.

File: src/Automatonymous/AutomatonymousStateMachine.cs [hunk 4]
```diff
@@ -1185,5 +1358,23 @@ public void Declare(object stateMachine)
...
+            StateMachineModifier<TInstance> builder = new InternalStateMachineModifier<TInstance>(this);
+            modifier(builder);
+            builder.Apply();
+
+            return this;
+        }
+
+        private class BuilderStateMachine : AutomatonymousStateMachine<TInstance> { }
+
+        public static AutomatonymousStateMachine<TInstance> Build(Action<StateMachineModifier<TInstance>> modifier)
```
Comment (phatboyg) [inline src/Automatonymous/AutomatonymousStateMachine.cs:1373]: I'm not sure I like this anchor, it feels like it should be elsewhere to align with the factory method approach used by all the MassTransit related project. Or maybe this is a sensible place, I can't think of another anchor and having a non-generic version of ASM would be annoying.

Maybe change Build to Create() or CreateUsingBuilder()

File: src/Automatonymous/Builders/InternalStateMachineEventActivitiesBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,174 @@
+﻿using Automatonymous.Binders;
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+
+namespace Automatonymous
+{
+    internal class InternalStateMachineEventActivitiesBuilder<TInstance> : StateMachineEventActivitiesBuilder<TInstance>
```
Comment (phatboyg) [inline src/Automatonymous/Builders/InternalStateMachineEventActivitiesBuilder.cs:8]: They're internal, sure, but I'd like to see all of the builder interfaces in a namespace, like Configuration/Builder or something, to avoid cluttering the root namespace.

File: src/Automatonymous/States/StateMachineState.cs [hunk 6]
```diff
@@ -216,9 +214,15 @@ public void AddSubstate(State<TInstance> subState)
             if (_name.Equals(subState.Name))
                 throw new ArgumentException("A state cannot be a substate of itself", nameof(subState));
 
+            subState.SetSuperState(this);
             _subStates.Add(subState);
         }
 
+        public void SetSuperState(State<TInstance> superState)
```
Comment (phatboyg) [inline src/Automatonymous/States/StateMachineState.cs:221]: Yeah, I still don't see a reason for this to be modifiable but there are a lot of changes. Also, this would be another internal method I think.

**Review by phatboyg — CHANGES_REQUESTED** [review]
So, this is pretty awesome, really great work. A few minor comments and ultimately this will make a nice feature for building state machines outside of the ctor.

**Comment by phatboyg** [comment]
Is your intention to make this usable with MassTransit as well, or just Automatonymous?

File: src/Automatonymous/AutomatonymousStateMachine.cs (hunk 3 above)
Comment (Jcouls29) [reply src/Automatonymous/AutomatonymousStateMachine.cs:707]: I'm open to suggestions here. The problem came in about defining sub-state when States were already defined. The approach could be changed so that the declaration of sub-state, creates the new state instead.

Right now, this is how to use it with the builder pattern
```csharp 
.Build(builder => builder
  .State("SubState", out State<TInstance> SubState)
  .State("SuperState", out State<TInstance> SuperState)
  .SubState(SubState, SuperState)
);
```

However, maybe this is a better approach:
```csharp
.Build(builder => builder
  .State("SuperState", out State SuperState)
  .SubState("SubState", SuperState, out State SubState)
);
```

This may be better, but would rather you make that call. Thoughts?

File: src/Automatonymous/AutomatonymousStateMachine.cs (hunk 4 above)
Comment (Jcouls29) [reply src/Automatonymous/AutomatonymousStateMachine.cs:1373]: Funny, I thought about `Create` first too. If you think that is appropriate, I'll rename it. As for a better anchor, I played with a couple different approaches including opening the builder up publicly to allow it directly, but it felt like too big of a change. However, creating a static class for `AutomatonymousStateMachine` with a single `Create<TInstance>` wouldn't be hard to do and might be better. How do you feel about this:

```csharp
public static class AutomatonymousStateMachine
{
  public static StateMachine<TInstance> Create<TInstance>(StateMachineModifier<TInstance> builder)
  {
    var machine = new BuilderStateMachine();
    machine.Modify(modifier);
    return machine;
  }

  private class BuilderStateMachine : AutomatonymousStateMachine<TInstance> { }
}
```

I can move the internal BuilderStateMachine here as well. I wouldn't inherit it from the current one. It would simply be a helper static class but externally wouldn't require any generic on the class itself. Just a thought.

File: src/Automatonymous/States/StateMachineState.cs (hunk 6 above)
Comment (Jcouls29) [reply src/Automatonymous/States/StateMachineState.cs:221]: I agree. It felt like a hack at the time. It has never felt right. I think based on the decision above (about how to declare SubState) will change whether I can remove it altogether or simply make it internal. Let me know your thoughts on that previous comment.

**Comment by Jcouls29** [comment]
In all honesty, I'm new to Automatonymous itself and haven't really touched MassTransit. I wouldn't feel comfortable doing any work outside of this library until I understand it more. If you're saying that this change will require some changes over there, I can definitely help modify to the best of my ability there as well. I had no intention of using MassTransit or modifying it.

File: src/Automatonymous/AutomatonymousStateMachine.cs (hunk 1 above)
Comment (Jcouls29) [reply src/Automatonymous/AutomatonymousStateMachine.cs:254]: I believe this is re-using more the code together. Although I had to add another method in the process.

File: src/Automatonymous/AutomatonymousStateMachine.cs (hunk 3 above)
Comment (Jcouls29) [reply src/Automatonymous/AutomatonymousStateMachine.cs:691]: @phatboyg I took a look at this again. I already had this other method created. I just removed the other way of creating sub-states from the builder pattern. Not sure it adds enough value. Let me know if you feel differently.

File: src/Automatonymous/States/StateMachineState.cs (hunk 6 above)
Comment (Jcouls29) [reply src/Automatonymous/States/StateMachineState.cs]: @phatboyg Removed this for now. The builder method `SubState(State<TInstance>, State<TInstance>)` was the method that required it. I removed this. I can see this causing inconsistencies anyways.

**Comment by phatboyg** [comment]
No need to modify MassTransit, so long as Automatonymous remains fully compatible with MassTransit after adding the builder features.

**Comment by Jcouls29** [comment]
@phatboyg Sounds good. I tried to make these changes "additive" in nature and not cause any issues with how folks use it now. Just let me know about creating the static ASM class. If you'd like me to add that and alter the tests to use it, I can. Otherwise, I have completed all the other changes requested.

**Comment by phatboyg** [comment]
I'll pull in your changes tonight and look them over again. Thanks!

**Comment by phatboyg** [comment]
Still on my todo list...

**Comment by phatboyg** [comment]
This has been merged, I changed Create to New since Create is pretty commonly used as an Event name.
